### ISSUE TITLE
`--explain` should disambiguate nonexistent from underdocumented error codes

### ISSUE THREAD

**Issue by zackmdavis** [rust-lang/rust#44710]
The `--explain EXXX` functionality (does anyone still uses this now that it's [no longer advertised so hard](https://github.com/rust-lang/rust/commit/fad4f32c3#diff-4d5b8cff35812b800ff77376382671a7L781)?) gives the same "no extended information" message for real error codes that don't have a "long" explanation written, and nonexistent codes. These messages should be disambiguated (as "no extended information for E0XXX" _vs_. "E0XXX does not exist" or similar).

```
$ rustc --explain E0489 # "type/lifetime parameter not in scope here"
error: no extended information for E0489

$ rustc --explain E9999nonexisting
error: no extended information for E9999nonexisting
```

-----

I _thought_ this would be straightforward to implement: we [_do_ keep track of codes without explanations](https://github.com/rust-lang/rust/blob/f60bc3ac/src/libsyntax/diagnostics/plugin.rs#L100) (which get successfully [plumbed through to the error index](https://github.com/rust-lang/rust/blob/f60bc3ac0c20f7b3/src/tools/error_index_generator/main.rs#L104), although the [CSS hides them](https://github.com/rust-lang/rust/blob/f60bc3ac0c20f7b3/src/tools/error_index_generator/main.rs#L68-L69)), only to [throw that information away](https://github.com/rust-lang/rust/blob/f60bc3ac/src/libsyntax/diagnostics/plugin.rs#L198-L199) when macroexpanding the arrays [that ultimately stock the `--explain` functionality](https://github.com/rust-lang/rust/blob/f60bc3ac0c20/src/librustc_driver/lib.rs#L1329-L1342). So this _should_, you would _think_, just be a matter of editing [`expand_build_diagnostic_array`](https://github.com/rust-lang/rust/blob/f60bc3ac/src/libsyntax/diagnostics/plugin.rs#L159) to give us a `[(&str, Option<&str>)]` of code–maybe-explanation pairs instead of a `[(&str, &str)]` of code–explanation pairs, editing [`errors::Registry`](https://github.com/rust-lang/rust/blob/f60bc3ac0c20f7b3ab85df1e5e2e3217a7a95630/src/librustc_errors/registry.rs) reflect that, and making the [driver](https://github.com/rust-lang/rust/blob/f60bc3ac0c20f7b3ab85df1e5e2e3217a7a95630/src/librustc_driver/lib.rs#L489) care.

But when I actually do this (zackmdavis/rust@048dde26e) ... my changes to `expand_build_diagnostic_array` seem to have absolutely no effect (!?!); we still get a `[(&str, &str)]` even though I _thought_ my code pretty clearly said otherwise, on more than one day that I've looked at this. I hereby give up in despair, bafflement, and tears.

This issue has been assigned to @jakevossen5 via [this comment](https://github.com/rust-lang/rust/issues/44710#issuecomment-588446206).

**Comment by Mark-Simulacrum** [comment]
The diagnostic registration is "just" a macro_rules macro now so I suspect this is pretty simple to implement per the suggested path in the issue description; I'm going to mark as E-easy and E-mentor with myself as the mentor.

You'll want to edit the `DIAGNOSTICS` array definitions [here](https://github.com/rust-lang/rust/blob/master/src/libsyntax/diagnostics/macros.rs#L178) to have the type `(&str, Option<&str>)` or so and add in the short diagnostics.

Once that's done, if you run `./x.py check` you'll likely hit an error in the use site of `DIAGNOSTICS` from [all the crates](https://github.com/rust-lang/rust/blob/master/src/librustc_interface/util.rs#L44). Fixing that'll mostly involve fixing up the diagnostics registry in [`librustc_errors`](https://github.com/rust-lang/rust/blob/master/src/librustc_errors/registry.rs); I would store the long diagnostics and short diagnostics separately, adding a field like `short_descriptions: FxHashSet<&'static str>`.

Once this is done, there might be a few more type errors to clean up but that should mostly be a matter of making types match up I would hope.

Let me know if you have any questions; here on this issue or feel free to open up a PR with some initial work and I can try and provide feedback there.

**Comment by Mark-Simulacrum** [comment]
One additional thing that we should probably add as part of this work is an error if the same diagnostic code is registered twice (cc https://github.com/rust-lang/rust/issues/64426). I imagine this would mostly be a matter of inserting all the codes into an FxHashSet and panicking if there's one inserted twice; an ICE here is fine since it'll definitely be caught in CI.

**Comment by jakevossen5** [comment]
Hey, 

If this still needs fixed, I would love to work on this. I will probably be reaching out before to long for some guidance. 

@rustbot claim

**Comment by jakevossen5** [comment]
Shoot not sure if I read what was happening in this thread all the way. @zackmdavis are you working on this already?

**Comment by zackmdavis** [comment]
@jakevossen5 I'm not; go for it! (Although ... I thought I remembered seeing an issue recently where people were discussing moving away from the numeric error codes, but I can't find it. @estebank, was that a real thing?)

**Comment by jakevossen5** [comment]
Maybe you are referring to this conversation on Zulip? https://rust-lang.zulipchat.com/#narrow/stream/147480-t-compiler.2Fwg-diagnostics/topic/get.20rid.20of.20error.20codes
